BITSWAP.md

```diff
@@ -196,6 +197,94 @@ message Message {
+      1. For each of the items that S sends back Blocks or BlockPresences for they may append any `tokens` they want. Recommended tokens include those that would inform C how they could convince S to give them access to the blocks they want.
+    2. If C sends S a Have request for data S does not have (or has but is not willing to tell C it has) and C has requested for DontHave responses then S should respond with DontHave
+    3. S may choose to include the number of bytes that are pending to be sent to C in the response message
+    4. If C asked for a block that S has but it is bigger than the maximum block size it should return `BlockTooBig`
+3. When C no longer needs a block it previously asked for it should send a Cancel message for that request to any peers that have not already responded about that particular block. It should particularly send Cancel messages for Block requests (as opposed to Have requests) that have not yet been answered.
+
+### Tokens
+
+The major change in this protocol version is the introduction of the ability to pass tokens along with requests and responses. A token is defined as `<multicode><data>` where the multicode is an identifier in the multicodec table, and the data is token-specific data associated with that code.
+
+Users who require additional codes for their new token formats should do one of:
+- Register their code in the table
+- For non-deployed testing purposes only - use a code in the application reserved range of the code table
+   - Note: if codes in the application reserved range will not be reservable in the code table which means that the code may conflict with another one used in the ecosystem which could cause application problems on collision. It is high recommended to not use these codes outside of testing or early development.
+
+To save space within the message the list of tokens used within the message are declared within the top level message and all other references to tokens are instead to the indices within the token list in the top level message.
+
+### Wire Format
+
+```
+message Message {
+
+  message Wantlist {
+    enum WantType {
+      Block = 0;
+      Have = 1;
+    }
+
+    message Entry {
+      bytes block = 1; // CID of the block
+      int32 priority = 2;	// the priority (normalized). default to 1
+      bool cancel = 3; // whether this revokes an entry
+      WantType wantType = 4; // Note: defaults to enum 0, ie Block
+      bool sendDontHave = 5; // Note: defaults to false
+      repeated tokens int32 = 6; // the indices of the tokens in the token list
+		}
+
+    repeated Entry entries = 1;	// a list of wantlist entries
+    bool full = 2; // whether this is the full wantlist. default to false
+  }
+  message Block {
+    bytes prefix = 1; // CID prefix (all of the CID components except for the digest of the multihash)
+    bytes data = 2;
+    repeated tokens int32 = 3; // the indices of the tokens in the token list
```

Comment:
from @ianopolous 
>  I think the simplest thing is just to avoid the field indices we've used

How important is this for you? There are two ways to handle this:
1. Skip field numbers (i.e. there is no field 3 defined)
2. We can use whatever numbers we want and since it's a different protocol version `/ipfs/bitswap/1.2.0` Block field 3 can be different from `/ipfs/bitswap/1.3.0` Block field 3.
    - This requires some custom code on your end at least until people upgrade to the new version. Not sure how annoying this is for you all though (e.g. if nodes upgrade quickly then it's probably not so bad since it only has to hang around for a little while)

Reply:
I'm just trying to think what our upgrade path with be like. If we skip the field numbers (1) then we can just trivially merge the protobuf and keep bitswap's single protobuf definition across all versions. That seems very desirable.

Reply:
Numbers skipped in https://github.com/ipfs/specs/pull/270/commits/acc0fe0f10016b5b397704293d15f72db25bae17